Level obstacles

Moneyl opened the pull request:
Changes:
- Level generation moved to `LevelGenerator.cs`. Makes a clear separation between level generation and level logic.
- Added time delta since last frame to `GameLogic.Update()`. Used for framerate independent code like player movement.
- All keys are now required to complete the end room and move into the next level. They're guaranteed to spawn somewhere in the level. Use the K key to add all keys to your inventory.
- Added special room types. Implemented by making classes that derive `Room` and implement custom per frame update functions. 
    - Windy rooms have constant wind that pushes the player around and periodically changes direction. 
    - Icy rooms have slippery floors and slower movement.
- Player movement is now based on velocity. This is so windy rooms can push the player around, and so players have to manage speed more carefully in icy rooms.
- Added floor spikes which damage the player when they collide. These only spawn in windy and icy rooms.
- Floor spikes are handled by a new class `LevelObject`. This is intended for interactive objects in the level that can't be picked up. Probably could merge this into `Item` and just have a "CanHold" flag.
- Fixed an issue with level cloning where the copy still referenced the original. I suspect this caused the bug where the wrong room would be rendered, then suddenly switch to the correct room. So far haven't been able to reproduce it again.

Moneyl committed:
Require keys to open end room

- Changed item spawning to only spawn the keys required for the end room once. They aren't spawned in the start room or end room, and each room can only have one of the keys.
- Added debug keybind "K" that adds all keys required by the end room to the players inventory.

Moneyl committed:
Use List<T> extension in Level

This is a bit clearer to read than the previous code since the function returns a bool when the list doesn't have the item instead of returning null.

Moneyl committed:
Fix previous commit

I wasn't careful enough while converting the code to use the extension method. Now works.

Moneyl committed:
Fix level gen bug

This bug was very rare. Sometimes a level would be generated with no connection from the start room to the end room. This was a bug in the pruning code. This was fixed by checking that a start-end room connection would still exist if a connection was pruned.

Moneyl committed:
Pass frame delta time to logic update functions

Needed to make object movement framerate independent. Specifically being used for player movement currently since it's being changed to use velocity to update position each frame in the next commit.

Moneyl committed:
Velocity based player movement

This change is being made so windy rooms can push the player around.

Moneyl committed:
Windy room obstacle

Some rooms are now windy. A wind sound effect plays  and the player is constantly pushed in the direction the wind is blowing. This isn't a big deal yet, but soon other hazards like spikes will be added that can harm the player is they run into them.

Moneyl committed:
Merge branch 'main' into LevelObstacles--10-27-2021

Moneyl committed:
Level gen code improvements

- Use Random.NextFloat() extension method min/max arguments instead of manually doing the range conversion. Much easier to read.
- Cleaned and commented secondary branching method (line 197 of Level.cs). A bit easier to understand.

Moneyl committed:
Level code cleanup

- Level generation code moved `LevelGenerator.cs`. This way level generation and level logic code are separate. Easier to understand and to find code.
- Different room types/obstacles are implemented by creating a new class which inherits `Room`. See Level.cs line 211.
- Added OnEnter() and OnExit() to Room as an easier way to start/stop room specific sound effects like wind.
- The player instance is now owned by GameLogic instead of Level. The player isn't recreated each time they beat a level anymore. This makes more sense and fixes inventory getting wiped every time a level is beat.
- Score is now preserved when completing a level.
- Keys are correctly removed from the players inventory when completing the end room.

Moneyl committed:
Add floor spike hazard

- Special rooms now spawn with 3-10 floor spikes. The player takes 1 damage if they run into them. There's a grace period before they can take damage again. That way there's some time to escape.
- The spikes are implemented by making a new class that inherits `LevelObject`. Other hazards like floor traps will be handled the same way.
- If they players health == 0, they die and the level is regenerated
- Added sound effects when the player dies, and when they hit a spike.

Moneyl committed:
Icy room type

- Adds icy rooms. Movement in these rooms is slower and the floor has much less friction so the player slides around. If they move too quickly they risk accidentally running into hazards.
- These rooms have an icy floor texture

Moneyl committed:
Item use sound effect

Moneyl committed:
Level cloning improvements

This fixes an oversight in Level.Clone() that meant that Rooms in cloned levels still referenced rooms in the original level instead of reference clones. So far this seems to have fixed a bug where the wrong room would be renderered. I tested it on seeds I noticed the issue on previously and it was no longer there. The bug was inconsistent so it's hard to say with complete certainty that it's fixed, will need to keep a look out while testing the game.

Moneyl committed:
Periodically change wind direction in windy rooms

Makes things a bit more interesting.

Moneyl committed:
Fix end room key console spam

If you're in the end room and don't have all keys the game writes the keys you have or don't have to the console. This commit fixes that getting spammed every frame. This is just a temporary fix, eventually the held keys should be shown on the UI.

Moneyl committed:
Move floor spike damage timer into Player.TryDamage()

Floor spikes have a timer that made the player safe for a short period of time after hitting a spike. This is required to some degree so the player has time to escape. This will be useful for other sources of damage too so it's been moved into Player.

Moneyl committed:
Fix level gen connection pruning bug

This bug broke connection pruning, resulting in much less interesting levels.

Moneyl committed:
Fix level gen unit test

trentv4 commented on src/logic/GameLogic.cs:
```diff
@@ -60,15 +61,39 @@ public class GameLogic {
+				Level = LevelGenerator.TryGenerateLevel();
+				Level.Player = Player;
+				Console.WriteLine($"Player died with a score of {Level.Score}!");
+				Renderer.EventQueue.Enqueue("LevelRegenerated"); //Signal to renderer to regenerate map scene
```
Later, this should be changed to a "PlayerDeath" so can I can implement a death screen. For now, good.

trentv4 commented on src/tests/LevelTests.cs:
```diff
@@ -8,11 +8,11 @@ namespace Project.Tests {
+			foreach (var room in level.Rooms) {
                 Assert.IsTrue(room.ConnectedRooms.Length >= 2);
             }
         }
```
I would add a test where it generates the level about 1000+ times to get a little extra security.

Moneyl replied:
Done.

trentv4 commented on src/logic/Level.cs:
```diff
@@ -493,21 +238,155 @@ public enum VisitedState {
+	/// <summary> Interactable object found in rooms. </summary>
+	public class LevelObject : ICloneable {
+		public Vector2 Position;
+		public readonly string TextureName;
```
This should be a reference to some sort of model loading after the model loading pull request is merged.

Moneyl committed:
Code review changes

- Moved floor friction from Player to Room
- Moved extension functions to single file
- Added test that generates many levels. The test only fails if LevelGenerator.TryGenerateLevel() fails. By default this requires level generation to fail with 1000 consecutive seeds. This signals a catostrophic failure since some aspect of the generation algo causes most seeds to fail.

Moneyl commented on src/logic/Player.cs:
```diff
@@ -12,6 +12,16 @@ public class Player : ICloneable {
+		public Vector2 Velocity;
+		public float MovementSpeed = DefaultMovementSpeed;
+		public readonly float MaxSpeed = 6.5f;
+		public float FloorFriction = DefaultFloorFriction;
```
Agreed, doesn't make sense to have that in `Player`. Moved to `Room` and it now passes the current room to `Player.Update()` so they can access it.

Moneyl commented on src/util/ListExtensions.cs:
```diff
@@ -0,0 +1,15 @@
+            return list.Find(match) != null;
+        }
+    }
+}
```
Done.
